File: backend/libs/nova_contracts/src/nova_contracts/universe.py
"""Stock list (D54): mirrors `frontend/packages/contracts/src/universe.ts`."""

from typing import Annotated

from pydantic import Field

from nova_contracts.common import Contract
from nova_contracts.market_data import IndexName

Symbol = Annotated[str, Field(pattern=r"^[A-Z0-9&-]{1,20}$")]
Text80 = Annotated[str, Field(max_length=80, pattern=r"\S")]


class UniverseEntryWrite(Contract):
    """Body of `POST /market-data/universe` and `PUT /market-data/universe/{symbol}`."""

    symbol: Symbol
    name: Text80
    sector: Text80
    indices: list[IndexName]


class UniverseEntry(UniverseEntryWrite):
    """`synced`: Kite knows the symbol (its instrument has a token), so it can be downloaded."""

    synced: bool


class InstrumentSyncResult(Contract):
    synced: list[Symbol]
    missing: list[Symbol]

File: frontend/packages/contracts/src/universe.ts
import { z } from "zod";
import { IndexNameSchema } from "./strategy";

/** Stock symbols as NSE writes them, e.g. `INFY`, `M&M`, `BAJAJ-AUTO`. */
export const UniverseSymbolSchema = z
  .string()
  .regex(/^[A-Z0-9&-]{1,20}$/, "1–20 capital letters, digits, & or -");

const text80 = (what: string) =>
  z.string().max(80, "At most 80 characters").regex(/\S/, `Enter the ${what}`);

/** Body of `POST /market-data/universe` and `PUT /market-data/universe/{symbol}` (D54). */
export const UniverseEntryWriteSchema = z.strictObject({
  symbol: UniverseSymbolSchema,
  name: text80("company name"),
  sector: text80("sector"),
  indices: z.array(IndexNameSchema),
});
export type UniverseEntryWrite = z.infer<typeof UniverseEntryWriteSchema>;

/** One stock of the list; `synced` = Kite knows it (it has an instrument token); `newListing` = a
 * sync added it after an earlier sync, e.g. an IPO (D56). */
export const UniverseEntrySchema = z.strictObject({
  ...UniverseEntryWriteSchema.shape,
  synced: z.boolean(),
  newListing: z.boolean(),
});
export type UniverseEntry = z.infer<typeof UniverseEntrySchema>;

/** One sector of the stock list and its stock count (`GET /market-data/universe/sectors`, D57). */
export const UniverseSectorSchema = z.strictObject({
  sector: z.string().min(1),
  count: z.number().int().min(1),
});
export type UniverseSector = z.infer<typeof UniverseSectorSchema>;
